src/crimson/net/ProtocolV2.cc

```diff
@@ -1825,7 +1825,9 @@ ceph::bufferlist ProtocolV2::do_sweep_messages(
 
     msg->encode(conn.features, 0);
 
-    ceph_assert(!msg->get_seq() && "message already has seq");
+    if (const auto seq = msg->get_seq(); seq) {
+      logger().info("{} already has seq={}, reassigning", *msg, seq);
+    }
```

Comment:
Under `!conn.policy.lossy` I think it is wrong to resend a message. The messages will be stored in `conn.sent` before it is acknowledged. The unacked messages need to be sent again in the exact order when network is broken. This part of history should not be modified.

Reuing a sent message has undesired side-effects:
- Reassigning the imbedded `seq` means the seq of the previous message in `conn.sent` can be changed. And the peer may see skipping sequences as a result;
- If user reuses the same message and modifies it, the sent history could be implicitly modified.

Under `conn.policy.lossy` I feels it is suspecious. The send process can be asynchronous (there is a `conn.out_q`, or if the underlying socket lives in a different core in the m:n model). If user really wants to reuse a message, they must be careful not to modify it because it may result in modifing the pending messages.

I think the best way is to copy the message if user want to resend it, and keep the `ceph_assert(!msg->get_seq())`. It may looks good to reuse a messange under lossy policy, but the correctness is now up to user not to modify them.

Reply:
@cyx1231st ack. will keep a copy instead.

Reply:
@rzarzynski @cyx1231st could you please help review #39798 ?